client/src/selectionRange.proposed.ts

```diff
@@ -0,0 +1,121 @@
+		super(client, SelectionRangeRequest.type);
+	}
+
+	public fillClientCapabilities(capabilites: ClientCapabilities): void {
+		let capability = ensure(ensure(capabilites, 'textDocument')!, 'selectionRange')!;
+		capability.dynamicRegistration = true;
+	}
+
+	public initialize(capabilities: ServerCapabilities, documentSelector: DocumentSelector): void {
+		if (!capabilities.selectionRangeProvider) {
+			return;
+		}
+
+		const implCapabilities = capabilities.selectionRangeProvider as TextDocumentRegistrationOptions & StaticRegistrationOptions & SelectionRangeProviderOptions;
+		const id = Is.string(implCapabilities.id) && implCapabilities.id.length > 0 ? implCapabilities.id : UUID.generateUuid();
+		const selector = implCapabilities.documentSelector || documentSelector;
+		if (selector) {
+			this.register(this.messages, {
+				id,
+				registerOptions: Object.assign({}, { documentSelector: selector })
+			});
+		}
+	}
+
+	protected registerLanguageProvider(options: TextDocumentRegistrationOptions): Disposable {
+		let client = this._client;
+		let provideSelectionRanges: ProvideSelectionRangeSignature = (document, position, token) => {
+			const requestParams: TextDocumentPositionParams = {
+				textDocument: client.code2ProtocolConverter.asTextDocumentIdentifier(document),
+				position: client.code2ProtocolConverter.asPosition(position),
+			};
+			return client.sendRequest(SelectionRangeRequest.type, requestParams, token).then(
+				(ranges) => isArray(ranges)
+					? ranges.map((r: Range) => client.protocol2CodeConverter.asRange(r))
+					: null,
+				(error: any) => {
+					client.logFailedRequest(SelectionRangeRequest.type, error);
+					return Promise.resolve(null);
+				}
+			);
+		};
+		let middleware = client.clientOptions.middleware!;
+		return Languages.registerSelectionRangeProvider(options.documentSelector!, {
+			provideSelectionRanges(document: TextDocument, position: VPosition, token: CancellationToken): ProviderResult<VRange[]> {
+				return middleware.provideSelectionRanges
+					? middleware.provideSelectionRanges(document, position, token, provideSelectionRanges)
+					: provideSelectionRanges(document, position, token);
+
+			}
+		});
+	}
+}
+
+declare module 'vscode' {
+
+	//#region Joh - selection range provider
+
+	export interface SelectionRangeProvider {
+		/**
+		 * Provide selection ranges starting at a given position. The first range must [contain](#Range.contains)
+		 * position and subsequent ranges must contain the previous range.
+		 * @param document
+		 * @param position
+		 * @param token
+		 */
+		provideSelectionRanges(document: TextDocument, position: Position, token: CancellationToken): ProviderResult<Range[]>;
+	}
+
+	export namespace languages {
+		export function registerSelectionRangeProvider(selector: DocumentSelector, provider: SelectionRangeProvider): Disposable;
+	}
+
+	//#endregion
+
+	//#region Joh - read/write in chunks
+
+	export interface FileSystemProvider {
```

Comment:
What is this interface for?

Reply:
Sloppy copy paste :sweat_smile:, removed!